tensorflow/lite/kernels/activations.cc

```diff
@@ -576,42 +577,72 @@ TfLiteStatus SoftmaxPrepare(TfLiteContext* context, TfLiteNode* node) {
   TF_LITE_ENSURE_OK(context, GetInputSafe(context, node, 0, &input));
   TfLiteTensor* output;
   TF_LITE_ENSURE_OK(context, GetOutputSafe(context, node, 0, &output));
-  if (output->type == kTfLiteInt16) {
-    TF_LITE_ENSURE(context, input->type == kTfLiteInt8 ||
-                                input->type == kTfLiteUInt8 ||
-                                input->type == kTfLiteInt16);
-  } else {
-    TF_LITE_ENSURE_TYPES_EQ(context, input->type, output->type);
-  }
 
   TF_LITE_ENSURE(context, NumDimensions(input) >= 1);
 
+  if (input->type == kTfLiteUInt8 || input->type == kTfLiteInt8 ||
+      input->type == kTfLiteInt16) {
+    if (input->type == kTfLiteInt16) {
+      TF_LITE_ENSURE_TYPES_EQ(context, output->type, kTfLiteInt16);
+      TF_LITE_ENSURE_EQ(context, output->params.zero_point, 0);
+      TF_LITE_ENSURE_NEAR(context, output->params.scale, 1.f / 32768,
```

Comment:
why this one is 32768 but ln-595 is 65536?

the softmax should output between [0,1], so all of them should be 65536 right?

Reply:
The zero-point is null for this one contrary to the 65536 one which has a zero-point of -32768.

Reply:
that confused me even more, I can understand that int16 ranges from -32768 to 32767, and if we minus the zero_point (-32768) and times the output scale, we got a range of [0, 1].

but it seems for this case, we will have a range of [-1, 1]?

Reply:
It's mainly because all int16 tensors have a null zero-point due to hardware constraints (constraint is checked also in the other kernels). With a null zero-point we can effectively represent a [-1, 1] range but it isn't a problem in itself as we need to preserve the null zero-point for int16 kernels.

I'm not sure about the (u)int8 -> int16 case and why it's an exception to these constraints, I used the same check as TFLite Micro but someone it would be better double check if it makes sense. How do we even generate models with such operators?

Reply:
but softmax should have [0, 1] outputs right?

Reply:
Yes but even if the int16 Tensor theoretically allows [-1, 1], the output will still be [0, 1], all int16 quantized values in the output tensor will just be >= 0.